### PR TITLE
switcher as web component

### PR DESCRIPTION
**Pull request by dgrammatiko**
Pull Request for Issue # .

### Summary of Changes
Switcher is now a true web component.

### Why?
- Truly decouple css of the component from any css framework
- Allow the component to disconnect/reconnect without breaking (eg try resizing the article edit and you'll figure out what's fixed there)
- support for web components is looking great now (Chrome, Opera, Safari, FF(next ver)) only MS is missing, but should come before J4
- Prepare for the future, eg support all the frameworks out there: check https://custom-elements-everywhere.com

### Testing Instructions
@ciar4n can you check that everything is ok here (eg edit an article and check that the resize is working ok now and also that save is not messed up)

### Expected result
Not broken

### Actual result
Broken

### Documentation Changes Required

@wilsonge you'll need to merge this

### CODE REVIEW

**Comment by ciar4n** [comment]
Very nice. I am unable to find any issues. Working well with shadow dom on Chrome. Fallback back to CE also working (Firefox).

**Comment by chmst** [comment]
The switcher works fine but I suggest to discuss if we need a visible label?
I think if there is no visible label at all, everyone understands what the switcher shows and what to do. Green (or dark for colorblind) = on. Light = off. Klick on the switcher to change th state.

Now the switcher for example shows green = yes. The label says "no".
Or the switcher is muted, but the label says "yes". This is confusing. 

Alternatives are:
The text additionally is vilible on the switcher itself, but this makes Problems with translated text.
Labels left and right: "yes" - buttons -"no". 
This is correct and everyone understands (but is is a challenge for designers) because the swichers the are not in line with input fields.
.

**Comment by dgrammatiko** [comment]
> but this makes Problems with translated text.

That shouldn't be a problem as we're using properly translatable strings there

> but is is a challenge for designers

that will make a lot of people angry

**Comment by brianteeman** [comment]
Using colour alone to signify any meaning will be an a11y fail. If you can't see colours etc then it's a problem. The golden rule is always to have at least two visual indicators

**Comment by chmst** [comment]
@brianteeman in this case, the different intensity of colour could be sufficient. Because the colour has no meaning, a switcher could be blue or pink as well. The colour of the muted buttons or the label is an a11y issue, a contrast ratio of about 1.5 is too low. But this is easily changed in the css so I did not mention it.

**Comment by dgrammatiko** [comment]
Reminder here, implement a solution (eg attribute inverted) for: https://github.com/joomla/joomla-cms/issues/20314

**Comment by Bakual** [comment]
I agree that the labels are confusing as it is.

I also think colors will not be enough. Especially when it comes to parameters where "enabled" is where you actually turn something off. We have this in J3 in several places where the button colors are inverted (eg red for 1 and green for 0). The most known example being the "Site Offline" parameter.
With the switcher, it will be green when the site is set offline, it should be red.

**Comment by chmst** [comment]
@Bakual green/red or other colours are an a11y issue. But your example shows very well why colours are wrong on switchers. 
See for example the issue tracker. 
switcher
If the label would be "Show Help?" Then the swicher would change the state on = yes/off = no. No red/green or green/blue  but different intensity of a colour - this is correct in the implementation by @dgrammatiko

**Comment by dgrammatiko** [comment]
Since people were unhappy with the encapsulation here is another approach:
screen shot 2018-05-21 at 23 28 25

PROS
- Still true web component
- uses regular file.css, lazy loaded (only once)
- overrides still work (eg /templates/mything/css...)

Cons
- css and js paths are predefined and correlated ( eg something/css/file.css and something/js/file.js)

@ciar4n @wilsonge

**Comment by wilsonge** [comment]
So after discussion in skype this is approved pending the writing up of CSS variables (and a issue in the tracker to discuss ie11 effects)

File: build/webcomponents/js/field-switcher/field-switcher.js [hunk 1]
```diff
@@ -5,6 +5,16 @@
 		SPACE: 32,
 	};
 
+	const template = document.createElement('template');
+	const path = document.currentScript.src;
+	template.innerHTML = `<link href="${path.replace('/js/', '/css/').replace('.js', '.css').replace('-es5', '')}" rel="stylesheet" type="text/css"></link>
```
Comment (C-Lodder) [inline build/webcomponents/js/field-switcher/field-switcher.js]: `type="text/css"` << not needed

Comment (dgrammatiko) [reply build/webcomponents/js/field-switcher/field-switcher.js]: unfortunately `document.currentScript` doesn't work with IE, so we have to encapsulate the css here. To ease the pain for newcomers we will also provide a PWA where someone will be able to tweak the colours or the scss and get an installable plugin that will create overrides to the current template and then self destruct. (or something like that...)

Comment (C-Lodder) [reply build/webcomponents/js/field-switcher/field-switcher.js]: Im simply referring to the `type` attribute which is not required in this case when using HTML5

**Pull request closed without merging**